builtin/providers/aws/resource_aws_lambda_function.go

```diff
@@ -297,19 +297,34 @@ func resourceAwsLambdaFunctionUpdate(d *schema.ResourceData, meta interface{}) e
 	}
 
 	codeUpdate := false
-	if v, ok := d.GetOk("filename"); ok && d.HasChange("source_code_hash") {
-		file, err := loadFileContent(v.(string))
-		if err != nil {
-			return fmt.Errorf("Unable to load %q: %s", v.(string), err)
+	if v, ok := d.GetOk("filename"); ok {
+		if d.HasChange("source_code_hash") {
+			file, err := loadFileContent(v.(string))
+			if err != nil {
+				return fmt.Errorf("Unable to load %q: %s", v.(string), err)
+			}
+			codeReq.ZipFile = file
+			codeUpdate = true
+		}
+	} else {
+		s3Bucket, bucketOk := d.GetOk("s3_bucket")
+		s3Key, keyOk := d.GetOk("s3_key")
+		s3ObjectVersion, versionOk := d.GetOk("s3_object_version")
+
+		if bucketOk && keyOk {
+			if d.HasChange("s3_bucket") || d.HasChange("s3_key") {
+				codeReq.S3Bucket = aws.String(s3Bucket.(string))
+				codeReq.S3Key = aws.String(s3Key.(string))
```

Comment:
Do you think we should set these even if they haven't changed?

I'm thinking about the case where only `s3_object_version` has changed and `s3_bucket` and `s3_key` are still the same... in that case, I think we'd try to send a request with an empty `S3Bucket` and `S3Key`, right?

I think it might work for us to separate the "has it changed?" logic from the construction of the object. So as pseudo-code it would be like:

```go
if any of s3_bucket, s3_key, s3_object_version, filename or source_code_hash have changed {
    codeReq = &lambda.UpdateFunctionCodeInput{
        FunctionName: aws.String(d.Id()),
    }
    if filename is set {
        codeReq.ZipFile = loadFileContent(v.(string))
    }
    else {
        codeReq.S3Bucket = d.Get("s3_bucket")
        codeReq.S3Key = d.Get("s3_key")
        if version is set {
            codeReq.S3ObjectVersion = d.Get("s3_object_Version")
        }
    }
    call UpdateFunctionCode with codeReq, handle errors
    call setPartial for all the code-related attributes
}
```

Do you think that makes sense? I think this way we'll always send S3Bucket and S3Key if there have been *any* changes to S3 attributes.